src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -706,11 +707,45 @@ public void MakeReadOnly()
         }
 
         /// <summary>
-        /// Configures the instance for use by the JsonSerializer APIs.
+        /// Marks the current instance as read-only preventing any further user modification.
+        /// </summary>
+        /// <param name="populateMissingResolver">Populates unconfigured <see cref="TypeInfoResolver"/> properties with the reflection-based default.</param>
+        /// <exception cref="InvalidOperationException">
+        /// The instance does not specify a <see cref="TypeInfoResolver"/> setting. Thrown if <paramref name="populateMissingResolver"/> is <see langword="false"/>.
+        /// -OR-
+        /// The <see cref="JsonSerializer.IsReflectionEnabledByDefault"/> feature switch has been turned off.
+        /// </exception>
+        /// <remarks>
+        /// When <paramref name="populateMissingResolver"/> is set to <see langword="true" />, configures the instance following
+        /// the semantics of the <see cref="JsonSerializer"/> methods accepting <see cref="JsonSerializerOptions"/> parameters.
+        ///
+        /// This method is idempotent.
+        /// </remarks>
+        [RequiresUnreferencedCode("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires unreferenced code.")]
+        [RequiresDynamicCode("Populating unconfigured TypeInfoResolver properties with the reflection resolver requires runtime code generation.")]
+        public void MakeReadOnly(bool populateMissingResolver)
+        {
+            if (populateMissingResolver)
+            {
+                if (!_isConfiguredForJsonSerializer)
+                {
+                    ConfigureForJsonSerializer();
```

Comment:
ConfigureForJsonSerializer() will throw an exception when `JsonSerializer.IsReflectionEnabledByDefault == false && TypeInfoResolver == null` saying:

```
Reflection-based serialization has been disabled for this application. Either use the source generator APIs or explicitly configure the 'JsonSerializerOptions.TypeInfoResolver' property.
```

I don't think ASP.NET Core will be able to use this new API because of this. There are scenarios where we need to call `MakeReadOnly()` on a JsonSerializerOptions that doesn't have its TypeInfoResolver set yet. One example is in MVC:

https://github.com/dotnet/aspnetcore/issues/49360

MVC always creates a SystemTextJsonOutputFormatter up front without knowing if the app is actually going to use JSON serialization or not. When the app doesn't touch JSON, we shouldn't be throwing an exception at startup.

Reply:
It looks like aspnetcore was able to work around https://github.com/dotnet/runtime/issues/89830 since it owns its `JsonSerializerOptions` instances, so we don't need new APIs to unblock that case. 

The semantics of the new overload is oriented towards components accepting `JsonSerializerOptions` instances they don't own but still need to emulate the behavior of the `JsonSerializer` methods.